pyomo/gdp/plugins/partition_disjuncts.py

```diff
@@ -0,0 +1,850 @@
+        nonlinear_decomp['nonlinear_vars'].append(
+            tuple(v for v in EXPR.identify_variables(summand)))
+
+    return nonlinear_decomp
+
+def arbitrary_partition(disjunction, P):
+    """
+    Returns a valid partition into P sets of the variables that appear in
+    algebraic additively separable constraints in the Disjuncts in
+    'disjunction'. Note that this method may return an invalid partition
+    if the constraints are not additively separable!
+
+    Arguments:
+    ----------
+    disjunction : A Disjunction object for which the variable partition will be
+                 created.
+    P : An int, the number of partitions
+    """
+    # collect variables
+    v_set = ComponentSet()
+    for disj in disjunction.disjuncts:
+        v_set.update(get_vars_from_components(disj, Constraint,
+                                              descend_into=Block, active=True))
+    # assign them to partitions
+    partitions = []
+    V = len(v_set)
+    whole = floor(V/P)
+    for partition in range(V % P):
+        partitions.append(ComponentSet())
+        # add whole + 1 vars
+        for i in range(whole + 1):
+            partitions[partition].add(v_set.pop())
+    # for the rest, add whole vars
+    for partition in range(V % P, P):
+        partitions.append(ComponentSet())
+        for i in range(whole):
+            partitions[partition].add(v_set.pop())
+
+    return partitions
+
+def compute_optimal_bounds(expr, global_constraints, opt):
+    """
+    Returns a tuple (LB, UB) where LB and UB are the results of minimizing
+    and maximizing expr over the variable bounds and the constraints on the
+    global_constraints block. Note that if expr is nonlinear, even if one of
+    the min and max problems is convex, the other won't be!
+
+    Arguments:
+    ----------
+    expr : The subexpression whose bounds we will return
+    global_constraints : A Block which contains the global Constraints and Vars
+                         of the original model
+    opt : A configured SolverFactory to use to minimize and maximize expr over
+          the set defined by global_constraints. Note that if expr is nonlinear,
+          opt will need to be capable of optimizing nonconvex problems.
+    """
+    if opt is None:
+        raise GDP_Error("No solver was specified to optimize the "
+                        "subproblems for computing expression bounds! "
+                        "Please specify a configured solver in the "
+                        "'compute_bounds_solver' argument if using "
+                        "'compute_optimal_bounds.'")
+
+    # add temporary objective and calculate bounds
+    obj = Objective(expr=expr)
+    global_constraints.add_component(unique_component_name(global_constraints,
+                                                           "tmp_obj"), obj)
+    results = opt.solve(global_constraints)
+    if verify_successful_solve(results) is not NORMAL:
```

Comment:
It looks like we never use `verify_successful_solve` except to check that the answer is optimal... So we can switch all of those to `check_optimal_termination`. Should we just deprecate `verify_successful_solve`? It's slightly fancier since it distinguishes infeasible as well, but apparently we haven't cared so far.